schema/query.go

```diff
@@ -107,6 +121,56 @@ func validateQuery(q map[string]interface{}, validator Validator, parentKey stri
 	queries := Query{}
 	for key, exp := range q {
 		switch key {
+		case "$regex":
+			if parentKey == "" {
+				return nil, errors.New("$regex can't be at first level")
+			}
+
+			switch exp.(type) {
+			case string:
+				regex := exp.(string)
+				v, err := regexp.Compile(regex)
+				if err != nil {
+					return nil, fmt.Errorf("$regex: invalid regex: %v", err)
+				}
+
+				value := RegexValue{
+					RegValue:   v,
+					RegOptions: "",
+				}
+
+				queries = append(queries, Regex{Field: parentKey, Value: value})
+			case []interface{}:
+				regex := exp.([]interface{})
+
+				var (
+					v   *regexp.Regexp
+					err error
+				)
+				// Control regex options behavour.
+				// First check if we find any options of the regexp package
+				// Supported options (flags):
+				// i              case-insensitive (default false)
+				// m              multi-line mode: ^ and $ match begin/end line in addition to begin/end text (default false)
+				// s              let . match \n (default false)
+				// U              ungreedy: swap meaning of x* and x*?, x+ and x+?, etc (default false)
+				if strings.ContainsAny(regex[1].(string), "imsU") {
```

Comment:
I'm wondering if we need this notation after all. What about sticking to Go notation (i.e. embedded `(?flags)` prefix) and see later if we want to introduce more advanced notation.

In its current form, the code does not prevent you from using `(?flags)` with or without the array value format. So either we deal with this case or we just drop the array notation and just adopt Go notation.

Sorry @omani, I realize that I'm just telling you that your previous implementation was actually good enough :)